### PR TITLE
don't listen for TCP connections when `--port 0` is passed

### LINKED ISSUE #979: Support --port 0 to disable TCP listening
Please support `--port 0` to disable TCP listening and allow UDS only.

```
dragonfly[1207338]: F20230322 15:02:24.829391 1207338 dfly_main.cc:483] Check failed: GetFlag(FLAGS_port) > 0u (0 vs. 0)
dragonfly[1207338]: *** Check failure stack trace: ***
dragonfly[1207338]:     @     0x55ede7888c73  (unknown)
dragonfly[1207338]:     @     0x55ede7881437  (unknown)
dragonfly[1207338]:     @     0x55ede7882dbf  (unknown)
dragonfly[1207338]:     @     0x55ede73cd755  (unknown)
dragonfly[1207338]:     @     0x7f0003d8fd90  (unknown)
dragonfly[1207338]:     @     0x7f0003d8fe40  __libc_start_main
dragonfly[1207338]:     @     0x55ede73cf8ce  (unknown)
dragonfly[1207338]:     @              (nil)  (unknown)
dragonfly[1207338]: *** SIGABRT received at time=1679493744 on cpu 1 ***
dragonfly[1207338]: PC: @     0x7f0003dfca7c  (unknown)  pthread_kill
dragonfly.service: Main process exited, code=dumped, status=6/ABRT
dragonfly.service: Failed with result 'core-dump'.
```

### PR DESCRIPTION
**Pull request by talbii**
Closes #979 and #983.

When the flag `--port 0` is passed, we disable the following listeners:
- `main_listener` (default TCP listener, usually port 6379)
- `admin_listener` (optional listener, disabled nonetheless as it uses TCP)

And instead require a UNIX socket passed, to which we talk instead.

---

About #983: if we run Dragonfly as a service with `PrivateNetwork=true`, then it crashes instead with the error:

```
Could not open port 6379, error: Address family not supported by protocol
```

which makes sense, as the service blocks all *network* connections.

### CODE REVIEW

File: src/server/dfly_main.cc [hunk 1]
```diff
@@ -327,16 +331,26 @@ bool RunEngine(ProactorPool* pool, AcceptServer* acceptor) {
...
     }
+  } else if (tcp_disabled) {
+    LOG(ERROR)
+        << "Did not receive a unix socket to listen to, yet TCP listening is disabled. Exiting.";
+    exit(1);
   }
 
   std::uint16_t admin_port = GetFlag(FLAGS_admin_port);
-  if (admin_port != 0) {
+  if (admin_port != 0 && !tcp_disabled) {
```
Comment (romange) [inline src/server/dfly_main.cc:353]: I think it's fine to let `admin_port` allow TCP connectivity independently from `--port`. This allows maximum flexibility

File: src/server/dfly_main.cc [hunk 2]
```diff
@@ -479,7 +495,13 @@ Usage: dragonfly [FLAGS]
   sigemptyset(&act.sa_mask);
   sigaction(SIGILL, &act, nullptr);
 
-  CHECK_GT(GetFlag(FLAGS_port), 0u);
+  if (GetFlag(FLAGS_port) == 0u) {
+    auto usock = GetFlag(FLAGS_unixsocket);
+    CHECK_GT(usock.length(), 0u);
```
Comment (romange) [inline src/server/dfly_main.cc:500]: Instead of CHECK_GT, please write something "either --port or --unixsocket must be enabled, exiting..".

**Review by romange — COMMENTED** [review]
looks good

File: src/server/dfly_main.cc (hunk 2 above)
Comment (talbii) [reply src/server/dfly_main.cc:500]: This is handled later, inside `RunEngine`, should this logic move to main?

https://github.com/dragonflydb/dragonfly/blob/4657f993534a4c51d4c1945e08a23bee25c8f01f/src/server/dfly_main.cc#L347-L349

Comment (romange) [reply src/server/dfly_main.cc:500]: If `--port=0` is specified, will dragonfly check-fail here? if yes, then obviously handling later does not help here.

**Review by romange — COMMENTED** [review]
1. Have you checked it's working for all configurations?
2. Can you please squash all commits on your side into a nice single commit with a message detailing what you did? Pls include `Fixes #` in the message body, not the title.
pls Ask Roy/Adi in case you have trouble to "rebase -i" command for squashing.

**Comment by talbii** [comment]
sure, I have tested running Dragonfly inside and outside a service, with `--port 0`, with/without unix sockets, etc.

**Review by romange — APPROVED** [review]

**Pull request merged**